### PR TITLE
options: thinking config and effort level

### PR DESCRIPTION
**Pull request by Roasbeef**
Catches the Go SDK up to TS SDK v0.2.119 on the thinking/reasoning options. See `memory/catchup-v0.2.119/PLAN.md` §"PR 1".

## What changed

- `ThinkingConfig` type with `Adaptive` / `Enabled(budget)` / `Disabled` constructors, plus an optional `Display` hint (`summarized` | `omitted`).
- `EffortLevel` string type and constants `EffortLow`, `EffortMedium`, `EffortHigh`, `EffortXHigh`, `EffortMax`.
- `Options.Thinking` and `Options.Effort` fields with `WithThinking` / `WithEffort` functional-option helpers.
- Transport now encodes `--thinking`, `--max-thinking-tokens`, `--thinking-display`, and `--effort` to match the TS SDK's argv builder in `sdk.mjs`.
- `MaxThinkingTokens` left in place with a `Deprecated:` comment; `Thinking` takes precedence when both are set.

## Tests

`transport_test.go` gets table-driven coverage for:
- each `Thinking.Type` variant (adaptive / enabled-with-budget / disabled) emits the right argv;
- nil `Thinking` emits neither `--thinking` nor `--max-thinking-tokens`;
- `Thinking.Display` wires `--thinking-display` (and is omitted when `Type == "disabled"` or unset);
- each of the five `EffortLevel` constants round-trips; empty `Effort` emits nothing;
- `MaxThinkingTokens` deprecation path still works, including the zero-value → `--thinking disabled` edge case.

## References

- `sdk.d.ts` L1347–1381 (`Options.thinking`, `Options.effort`, deprecated `maxThinkingTokens`)
- `sdk.d.ts` L5250–5277 (`ThinkingConfig` union)
- `sdk.mjs` argv builder (switch on `thinking.type` and optional `thinking.display`)

### CODE REVIEW

File: transport_test.go [hunk 1]
```diff
@@ -644,6 +664,209 @@ func TestSubprocessTransportConnectArguments(t *testing.T) {
...
+			wantValue:  "4096",
+			absentFlag: "--thinking",
+		},
+		{
+			name:       "disabled",
+			thinking:   ThinkingDisabled(),
+			wantFlag:   "--thinking",
+			wantValue:  "disabled",
+			absentFlag: "--max-thinking-tokens",
+		},
```
Comment (Roasbeef) [inline transport_test.go:702]: This table covers the three constructors, but it misses the TS-valid `ThinkingEnabled` shape with no budget. `sdk.d.ts` makes `ThinkingEnabled.budgetTokens` optional (L5273-L5276), and `sdk.mjs` emits `--thinking adaptive` when `type === "enabled"` and `budgetTokens === void 0`. Since `ThinkingConfig` is exported and callers can build `&ThinkingConfig{Type: "enabled"}`, add a row for that branch.

File: transport_test.go [hunk 2]
```diff
@@ -644,6 +664,209 @@ func TestSubprocessTransportConnectArguments(t *testing.T) {
...
+	transport := NewSubprocessTransportWithRunner(runner, opts)
+
+	err := transport.Connect(context.Background())
+	require.NoError(t, err)
+	defer transport.Close()
+
+	assertArgAbsent(t, runner.StartArgs, "--effort")
+}
+
+func TestSubprocessTransportMaxThinkingTokensStillWorks(t *testing.T) {
```
Comment (Roasbeef) [inline transport_test.go:845]: The PR body says `Thinking` takes precedence over `MaxThinkingTokens`, and TS does that by deriving `thinkingConfig` from `thinking` before consulting `maxThinkingTokens` (`sdk.mjs` line 105). Please add a test with both fields set, for example `ThinkingDisabled()` plus `MaxThinkingTokens = 2048`, and assert only `--thinking disabled` is emitted. This protects the deprecation behavior this PR is promising.

**Review by Roasbeef — COMMENTED** [review]
Checked against TS SDK v0.2.119: `sdk.d.ts` L1347-L1381 and L5250-L5277, plus the minified `sdk.mjs` argv builder. The emitted flags match TS for adaptive / disabled / enabled-with-budget / enabled-without-budget, display omission on disabled, effort, and `maxThinkingTokens == 0`.

One API-doc cleanup to fold in: `MaxThinkingTokens` is marked deprecated on the field, but the exported `WithMaxThinkingTokens` helper still lacks a `Deprecated:` doc line. Go deprecation is attached per exported identifier, so callers using the functional option will not see the migration hint unless that helper comment is updated too.

Overall: flag encoding is right; add the edge-case/precedence tests and deprecate the helper so the compatibility path stays explicit.

File: transport_test.go (hunk 1 above)
Comment (Roasbeef) [reply transport_test.go:702]: Fixed in cd74ca7. Added the exported-struct case with Type "enabled" and nil BudgetTokens, asserting it emits --thinking adaptive and omits --max-thinking-tokens.

File: transport_test.go (hunk 2 above)
Comment (Roasbeef) [reply transport_test.go:845]: Fixed in cd74ca7. Added ThinkingDisabled plus MaxThinkingTokens=2048 coverage and asserted only --thinking disabled is emitted.

**Comment by Roasbeef** [comment]
Review body item addressed in cd74ca7: added a Deprecated: line to WithMaxThinkingTokens so callers using the functional option get the migration hint too.

**Comment by Roasbeef** [comment]
**Round 2 review — VERDICT: APPROVE**

Reviewed against @anthropic-ai/claude-agent-sdk@0.2.119. All round-1 gaps are addressed:

- `enabled` without `BudgetTokens` now tested, emits `--thinking adaptive` (matches TS SDK).
- `Thinking` precedence over `MaxThinkingTokens` pinned by test.
- `WithMaxThinkingTokens` has its own `Deprecated:` doc line.
- No new material issues.

CI green across all platforms. Approving and squash-merging.

_(Posted as comment instead of formal Approve event because the `gh` token is authed as the PR author; GitHub rejects self-approval. Verdict stands: approve+merge.)_
